File: packages/kotlin/src/main/kotlin/me/really/jose/ReallyMeJose.kt
// SPDX-FileCopyrightText: 2026 ReallyMe LLC
//
// SPDX-License-Identifier: MIT OR Apache-2.0

package me.really.jose

import com.google.protobuf.CodedInputStream
import com.google.protobuf.InvalidProtocolBufferException
import me.really.jose.v1.JoseJweDecryptRequest
import me.really.jose.v1.JoseJweDecryptResponse
import me.really.jose.v1.JoseJweEncryptRequest
import me.really.jose.v1.JoseJweEncryptResponse
import me.really.jose.v1.JoseJwtDecodeUnsignedRequest
import me.really.jose.v1.JoseJwtDecodeUnsignedResponse
import me.really.jose.v1.JoseJwtEncodeUnsignedRequest
import me.really.jose.v1.JoseJwtEncodeUnsignedResponse
import me.really.jose.v1.JoseJwtSignRequest
import me.really.jose.v1.JoseJwtSignResponse
import me.really.jose.v1.JoseJwtVerifyRequest
import me.really.jose.v1.JoseJwtVerifyResponse
import me.really.jose.v1.JoseOperationContractVersion
import me.really.jose.v1.JoseOperationRequest
import me.really.jose.v1.JoseOperationResponse
import me.really.jose.v1.JoseJwsSignRequest
import me.really.jose.v1.JoseJwsSignResponse
import me.really.jose.v1.JoseJwsVerifyRequest
import me.really.jose.v1.JoseJwsVerifyResponse


private const val MAX_PROTOBUF_RECURSION_DEPTH: Int = 32

/** Typed Kotlin/JVM facade over the canonical Rust JOSE operation contract. */
public object ReallyMeJose {
    @JvmStatic
    public fun signJws(
        algorithm: ReallyMeJoseSignatureAlgorithm,
        privateKey: ByteArray,
        payload: ByteArray,
    ): String = withOwned(privateKey, payload) { owned ->
        requireAggregate(owned[0].size, owned[1].size)
        val operation = JoseJwsSignRequest.newBuilder()
            .setAlgorithm(protoSignatureAlgorithm(algorithm))
            .setPrivateKey(wrap(owned[0]))
            .setPayload(wrap(owned[1]))
            .build()
        val response = execute(JoseOperationRequest.newBuilder().setJwsSign(operation).build())
        if (response.responseCase != JoseOperationResponse.ResponseCase.JWS_SIGN) malformed()
        val selected = response.jwsSign
        requireClean(selected)
        when (selected.outcomeCase) {
            JoseJwsSignResponse.OutcomeCase.RESULT -> compact(selected.result)
            JoseJwsSignResponse.OutcomeCase.ERROR -> throw sdkError(selected.error)
            else -> malformed()
        }
    }

    @JvmStatic
    public fun verifyJws(
        algorithm: ReallyMeJoseSignatureAlgorithm,
        compact: String,
        publicKey: ByteArray,
    ): Unit = withOwned(publicKey) { owned ->
        requireAggregate(utf8Length(compact), owned[0].size)
        val operation = JoseJwsVerifyRequest.newBuilder()
            .setAlgorithm(protoSignatureAlgorithm(algorithm))
            .setCompact(compact)
            .setPublicKey(wrap(owned[0]))
            .build()
        val response = execute(JoseOperationRequest.newBuilder().setJwsVerify(operation).build())
        if (response.responseCase != JoseOperationResponse.ResponseCase.JWS_VERIFY) malformed()
        val selected = response.jwsVerify
        requireClean(selected)
        when (selected.outcomeCase) {
            JoseJwsVerifyResponse.OutcomeCase.RESULT -> requireClean(selected.result)
            JoseJwsVerifyResponse.OutcomeCase.ERROR -> throw sdkError(selected.error)
            else -> malformed()
        }
    }

    @JvmStatic
    public fun encodeUnsignedJwt(claimsJson: ByteArray): String = withOwned(claimsJson) { owned ->
        requireAggregate(owned[0].size)
        val operation = JoseJwtEncodeUnsignedRequest.newBuilder()
            .setClaimsJson(wrap(owned[0]))
            .build()
        val response = execute(JoseOperationRequest.newBuilder().setJwtEncodeUnsigned(operation).build())
        if (response.responseCase != JoseOperationResponse.ResponseCase.JWT_ENCODE_UNSIGNED) malformed()
        val selected = response.jwtEncodeUnsigned
        requireClean(selected)
        when (selected.outcomeCase) {
            JoseJwtEncodeUnsignedResponse.OutcomeCase.RESULT -> compact(selected.result)
            JoseJwtEncodeUnsignedResponse.OutcomeCase.ERROR -> throw sdkError(selected.error)
            else -> malformed()
        }
    }

    @JvmStatic
    public fun decodeUnsignedJwt(compact: String): ByteArray {
        requireAggregate(utf8Length(compact))
        val operation = JoseJwtDecodeUnsignedRequest.newBuilder().setCompact(compact).build()
        val response = execute(JoseOperationRequest.newBuilder().setJwtDecodeUnsigned(operation).build())
        if (response.responseCase != JoseOperationResponse.ResponseCase.JWT_DECODE_UNSIGNED) malformed()
        val selected = response.jwtDecodeUnsigned
        requireClean(selected)
        return when (selected.outcomeCase) {
            JoseJwtDecodeUnsignedResponse.OutcomeCase.RESULT -> {
                requireClean(selected.result)
                selected.result.claimsJson.toByteArray()
            }
            JoseJwtDecodeUnsignedResponse.OutcomeCase.ERROR -> throw sdkError(selected.error)
            else -> malformed()
        }
    }

    @JvmStatic
    @JvmOverloads
    public fun signJwt(
        claimsJson: ByteArray,
        jwkJson: ByteArray,
        privateKey: ByteArray,
        type: String = "",
    ): String = withOwned(claimsJson, jwkJson, privateKey) { owned ->
        requireAggregate(owned[0].size, owned[1].size, owned[2].size, utf8Length(type))
        val operation = JoseJwtSignRequest.newBuilder()
            .setClaimsJson(wrap(owned[0]))
            .setJwkJson(wrap(owned[1]))
            .setPrivateKey(wrap(owned[2]))
            .setTyp(type)
            .build()
        val response = execute(JoseOperationRequest.newBuilder().setJwtSign(operation).build())
        if (response.responseCase != JoseOperationResponse.ResponseCase.JWT_SIGN) malformed()
        val selected = response.jwtSign
        requireClean(selected)
        when (selected.outcomeCase) {
            JoseJwtSignResponse.OutcomeCase.RESULT -> compact(selected.result)
            JoseJwtSignResponse.OutcomeCase.ERROR -> throw sdkError(selected.error)
            else -> malformed()
        }
    }

    @JvmStatic
    @JvmOverloads
    public fun verifyJwt(
        compact: String,
        jwkJson: ByteArray,
        publicKey: ByteArray,
        headerPolicy: ReallyMeJoseJwtHeaderPolicy? = null,
        temporalPolicy: ReallyMeJoseJwtTemporalPolicy? = null,
        signatureOnly: Boolean = false,
    ): ByteArray = withOwned(jwkJson, publicKey) { owned ->
        requireAggregate(utf8Length(compact), owned[0].size, owned[1].size)
        val builder = JoseJwtVerifyRequest.newBuilder()
            .setCompact(compact)
            .setJwkJson(wrap(owned[0]))
            .setPublicKey(wrap(owned[1]))
            .setSignatureOnly(signatureOnly)
        if (headerPolicy != null) builder.setHeaderPolicy(protoJwtHeaderPolicy(headerPolicy))
        if (temporalPolicy != null) builder.setTemporalPolicy(protoJwtTemporalPolicy(temporalPolicy))
        val response = execute(JoseOperationRequest.newBuilder().setJwtVerify(builder.build()).build())
        if (response.responseCase != JoseOperationResponse.ResponseCase.JWT_VERIFY) malformed()
        val selected = response.jwtVerify
        requireClean(selected)
        when (selected.outcomeCase) {
            JoseJwtVerifyResponse.OutcomeCase.RESULT -> {
                requireClean(selected.result)
                selected.result.claimsJson.toByteArray()
            }
            JoseJwtVerifyResponse.OutcomeCase.ERROR -> throw sdkError(selected.error)
            else -> malformed()
        }
    }

    @JvmStatic
    @JvmOverloads
    public fun encryptJwe(
        keyManagementAlgorithm: ReallyMeJoseJweKeyManagementAlgorithm,
        contentEncryptionAlgorithm: ReallyMeJoseJweContentEncryptionAlgorithm,
        key: ByteArray,
        plaintext: ByteArray,
        keyIdentifier: String = "",
        agreementPartyUInfo: ByteArray = ByteArray(0),
        agreementPartyVInfo: ByteArray = ByteArray(0),
        type: String = "",
        contentType: String = "",
    ): String = withOwned(key, plaintext, agreementPartyUInfo, agreementPartyVInfo) { owned ->
        requireAggregate(
            owned[0].size, owned[1].size, utf8Length(keyIdentifier), owned[2].size,
            owned[3].size, utf8Length(type), utf8Length(contentType),
        )
        val operation = JoseJweEncryptRequest.newBuilder()
            .setKeyManagementAlgorithm(protoKeyManagementAlgorithm(keyManagementAlgorithm))
            .setContentEncryptionAlgorithm(protoContentEncryptionAlgorithm(contentEncryptionAlgorithm))
            .setKey(wrap(owned[0]))
            .setPlaintext(wrap(owned[1]))
            .setKid(keyIdentifier)
            .setApu(wrap(owned[2]))
            .setApv(wrap(owned[3]))
            .setTyp(type)
            .setCty(contentType)
            .build()
        val response = execute(JoseOperationRequest.newBuilder().setJweEncrypt(operation).build())
        if (response.responseCase != JoseOperationResponse.ResponseCase.JWE_ENCRYPT) malformed()
        val selected = response.jweEncrypt
        requireClean(selected)
        when (selected.outcomeCase) {
            JoseJweEncryptResponse.OutcomeCase.RESULT -> compact(selected.result)
            JoseJweEncryptResponse.OutcomeCase.ERROR -> throw sdkError(selected.error)
            else -> malformed()
        }
    }

    @JvmStatic
    @JvmOverloads
    public fun decryptJwe(
        compact: String,
        keyManagementAlgorithm: ReallyMeJoseJweKeyManagementAlgorithm,
        contentEncryptionAlgorithm: ReallyMeJoseJweContentEncryptionAlgorithm,
        key: ByteArray,
        headerPolicy: ReallyMeJoseJweHeaderPolicy? = null,
    ): ByteArray = withOwned(key) { owned ->
        requireAggregate(utf8Length(compact), owned[0].size)
        val policyOwned = mutableListOf<ByteArray>()
        try {
            val builder = JoseJweDecryptRequest.newBuilder()
                .setCompact(compact)
                .setKeyManagementAlgorithm(protoKeyManagementAlgorithm(keyManagementAlgorithm))
                .setContentEncryptionAlgorithm(protoContentEncryptionAlgorithm(contentEncryptionAlgorithm))
                .setKey(wrap(owned[0]))
            if (headerPolicy != null) builder.setHeaderPolicy(protoJweHeaderPolicy(headerPolicy, policyOwned))
            val response = execute(JoseOperationRequest.newBuilder().setJweDecrypt(builder.build()).build())
            if (response.responseCase != JoseOperationResponse.ResponseCase.JWE_DECRYPT) malformed()
            val selected = response.jweDecrypt
            requireClean(selected)
            when (selected.outcomeCase) {
                JoseJweDecryptResponse.OutcomeCase.RESULT -> {
                    requireClean(selected.result)
                    selected.result.plaintext.toByteArray()
                }
                JoseJweDecryptResponse.OutcomeCase.ERROR -> throw sdkError(selected.error)
                else -> malformed()
            }
        } finally {
            policyOwned.forEach { it.fill(0) }
        }
    }

    /** Explicit wire API. The caller owns and must clear the returned array. */
    @JvmStatic
    public fun executeWireRequest(request: ByteArray): ByteArray = executeOwned(request, false)

    /** Explicit ProtoJSON wire API returning a canonical binary response. */
    @JvmStatic
    public fun executeWireJsonRequest(request: ByteArray): ByteArray = executeOwned(request, true)

    private fun execute(request: JoseOperationRequest): JoseOperationResponse {
        if (request.serializedSize > ReallyMeJoseRustNativeProvider.binaryRequestLimit()) {
            throw resourceLimit()
        }
        val requestBytes = request.toByteArray()
        val responseBytes = try {
            ReallyMeJoseRustNativeProvider.requireLoaded()
            ReallyMeJoseNative.executeOperationNative(requestBytes)
        } finally {
            requestBytes.fill(0)
        }
        try {
            val input = CodedInputStream.newInstance(responseBytes)
            input.setRecursionLimit(MAX_PROTOBUF_RECURSION_DEPTH)
            input.setSizeLimit(ReallyMeJoseRustNativeProvider.responseLimit())
            val response = JoseOperationResponse.parseFrom(input)
            if (
                response.contractVersion != JoseOperationContractVersion.JOSE_OPERATION_CONTRACT_VERSION_V1 ||
                response.reallyMeHasUnknownFieldsForValidation()
            ) malformed()
            if (response.responseCase == JoseOperationResponse.ResponseCase.BOUNDARY_ERROR) {
                throw sdkError(response.boundaryError)
            }
            return response
        } catch (_: InvalidProtocolBufferException) {
            malformed()
        } finally {
            responseBytes.fill(0)
        }
    }

    private fun executeOwned(request: ByteArray, json: Boolean): ByteArray {
        val limit = if (json) {
            ReallyMeJoseRustNativeProvider.jsonRequestLimit()
        } else {
            ReallyMeJoseRustNativeProvider.binaryRequestLimit()
        }
        if (request.size > limit) throw resourceLimit()
        val owned = request.copyOf()
        try {
            return if (json) {
                ReallyMeJoseNative.executeOperationJsonNative(owned)
            } else {
                ReallyMeJoseNative.executeOperationNative(owned)
            }
        } finally {
            owned.fill(0)
        }
    }

    private fun requireAggregate(vararg lengths: Int) {
        var aggregate = 0
        val maximum = ReallyMeJoseRustNativeProvider.binaryRequestLimit()
        for (length in lengths) {
            aggregate = try {
                Math.addExact(aggregate, length)
            } catch (_: ArithmeticException) {
                throw resourceLimit()
            }
            if (aggregate > maximum) throw resourceLimit()
        }
    }
}

File: packages/swift/Sources/ReallyMeJOSE/ReallyMeJOSE.swift
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

import Foundation
import ReallyMeJOSEProto
import SwiftProtobuf

/// Typed Swift facade over the canonical Rust JOSE operation contract.
public struct ReallyMeJOSE: Sendable {
  private let provider: any ReallyMeJOSENativeProvider

  public init(nativeLibrary: ReallyMeJOSENativeLibrary) throws {
    provider = try ReallyMeJOSERustProvider(library: nativeLibrary)
  }

  #if REALLYME_JOSE_LINKED_FFI
    public init() throws {
      provider = try ReallyMeJOSERustProvider()
    }
  #endif

  public func signJWS(
    algorithm: ReallyMeJOSESignatureAlgorithm,
    privateKey: [UInt8],
    payload: [UInt8]
  ) throws -> String {
    try requireAggregateInput([privateKey.count, payload.count])
    var operation = ReallyMeProtoJoseJwsSignRequest()
    operation.algorithm = protoSignatureAlgorithm(algorithm)
    operation.privateKey = Data(privateKey)
    operation.payload = Data(payload)
    var request = ReallyMeProtoJoseOperationRequest()
    request.operation = .jwsSign(operation)
    var response = try execute(&request)
    guard case .jwsSign(let selected)? = response.response,
      selected.unknownFields.data.isEmpty
    else {
      throw ReallyMeJOSEError.malformedProviderResponse
    }
    switch selected.outcome {
    case .result(let result):
      guard result.unknownFields.data.isEmpty else {
        throw ReallyMeJOSEError.malformedProviderResponse
      }
      response.response = nil
      return result.compact
    case .error(let error):
      throw try sdkError(error)
    case nil:
      throw ReallyMeJOSEError.malformedProviderResponse
    }
  }

  public func verifyJWS(
    algorithm: ReallyMeJOSESignatureAlgorithm,
    compact: String,
    publicKey: [UInt8]
  ) throws {
    try requireAggregateInput([compact.utf8.count, publicKey.count])
    var operation = ReallyMeProtoJoseJwsVerifyRequest()
    operation.algorithm = protoSignatureAlgorithm(algorithm)
    operation.compact = compact
    operation.publicKey = Data(publicKey)
    var request = ReallyMeProtoJoseOperationRequest()
    request.operation = .jwsVerify(operation)
    let response = try execute(&request)
    guard case .jwsVerify(let selected)? = response.response,
      selected.unknownFields.data.isEmpty
    else {
      throw ReallyMeJOSEError.malformedProviderResponse
    }
    switch selected.outcome {
    case .result(let result):
      guard result.unknownFields.data.isEmpty else {
        throw ReallyMeJOSEError.malformedProviderResponse
      }
    case .error(let error):
      throw try sdkError(error)
    case nil:
      throw ReallyMeJOSEError.malformedProviderResponse
    }
  }

  public func encodeUnsignedJWT(claimsJSON: [UInt8]) throws -> String {
    try requireAggregateInput([claimsJSON.count])
    var operation = ReallyMeProtoJoseJwtEncodeUnsignedRequest()
    operation.claimsJson = Data(claimsJSON)
    var request = ReallyMeProtoJoseOperationRequest()
    request.operation = .jwtEncodeUnsigned(operation)
    let response = try execute(&request)
    guard case .jwtEncodeUnsigned(let selected)? = response.response,
      selected.unknownFields.data.isEmpty
    else {
      throw ReallyMeJOSEError.malformedProviderResponse
    }
    switch selected.outcome {
    case .result(let result):
      guard result.unknownFields.data.isEmpty else {
        throw ReallyMeJOSEError.malformedProviderResponse
      }
      return result.compact
    case .error(let error): throw try sdkError(error)
    case nil: throw ReallyMeJOSEError.malformedProviderResponse
    }
  }

  public func decodeUnsignedJWT(_ compact: String) throws -> [UInt8] {
    try requireAggregateInput([compact.utf8.count])
    var operation = ReallyMeProtoJoseJwtDecodeUnsignedRequest()
    operation.compact = compact
    var request = ReallyMeProtoJoseOperationRequest()
    request.operation = .jwtDecodeUnsigned(operation)
    var response = try execute(&request)
    guard case .jwtDecodeUnsigned(var selected)? = response.response,
      selected.unknownFields.data.isEmpty
    else {
      throw ReallyMeJOSEError.malformedProviderResponse
    }
    response.response = nil
    switch selected.outcome {
    case .result(var result):
      guard result.unknownFields.data.isEmpty else {
        throw ReallyMeJOSEError.malformedProviderResponse
      }
      selected.outcome = nil
      defer { ReallyMeJOSEMemory.clearOwned(&result.claimsJson) }
      return [UInt8](result.claimsJson)
    case .error(let error): throw try sdkError(error)
    case nil: throw ReallyMeJOSEError.malformedProviderResponse
    }
  }

  public func signJWT(
    claimsJSON: [UInt8],
    jwkJSON: [UInt8],
    privateKey: [UInt8],
    type: String = ""
  ) throws -> String {
    try requireAggregateInput([claimsJSON.count, jwkJSON.count, privateKey.count, type.utf8.count])
    var operation = ReallyMeProtoJoseJwtSignRequest()
    operation.claimsJson = Data(claimsJSON)
    operation.jwkJson = Data(jwkJSON)
    operation.privateKey = Data(privateKey)
    operation.typ = type
    var request = ReallyMeProtoJoseOperationRequest()
    request.operation = .jwtSign(operation)
    let response = try execute(&request)
    guard case .jwtSign(let selected)? = response.response,
      selected.unknownFields.data.isEmpty
    else {
      throw ReallyMeJOSEError.malformedProviderResponse
    }
    switch selected.outcome {
    case .result(let result):
      guard result.unknownFields.data.isEmpty else {
        throw ReallyMeJOSEError.malformedProviderResponse
      }
      return result.compact
    case .error(let error): throw try sdkError(error)
    case nil: throw ReallyMeJOSEError.malformedProviderResponse
    }
  }

  public func verifyJWT(
    compact: String,
    jwkJSON: [UInt8],
    publicKey: [UInt8],
    headerPolicy: ReallyMeJOSEJWTHeaderPolicy? = nil,
    temporalPolicy: ReallyMeJOSEJWTTemporalPolicy? = nil,
    signatureOnly: Bool = false
  ) throws -> [UInt8] {
    try requireAggregateInput([compact.utf8.count, jwkJSON.count, publicKey.count])
    var operation = ReallyMeProtoJoseJwtVerifyRequest()
    operation.compact = compact
    operation.jwkJson = Data(jwkJSON)
    operation.publicKey = Data(publicKey)
    operation.signatureOnly = signatureOnly
    if let headerPolicy { operation.headerPolicy = protoJWTHeaderPolicy(headerPolicy) }
    if let temporalPolicy { operation.temporalPolicy = protoJWTTemporalPolicy(temporalPolicy) }
    var request = ReallyMeProtoJoseOperationRequest()
    request.operation = .jwtVerify(operation)
    var response = try execute(&request)
    guard case .jwtVerify(var selected)? = response.response,
      selected.unknownFields.data.isEmpty
    else {
      throw ReallyMeJOSEError.malformedProviderResponse
    }
    response.response = nil
    switch selected.outcome {
    case .result(var result):
      guard result.unknownFields.data.isEmpty else {
        throw ReallyMeJOSEError.malformedProviderResponse
      }
      selected.outcome = nil
      defer { ReallyMeJOSEMemory.clearOwned(&result.claimsJson) }
      return [UInt8](result.claimsJson)
    case .error(let error): throw try sdkError(error)
    case nil: throw ReallyMeJOSEError.malformedProviderResponse
    }
  }

  public func encryptJWE(
    keyManagementAlgorithm: ReallyMeJOSEJWEKeyManagementAlgorithm,
    contentEncryptionAlgorithm: ReallyMeJOSEJWEContentEncryptionAlgorithm,
    key: [UInt8],
    plaintext: [UInt8],
    keyIdentifier: String = "",
    agreementPartyUInfo: [UInt8] = [],
    agreementPartyVInfo: [UInt8] = [],
    type: String = "",
    contentType: String = ""
  ) throws -> String {
    try requireAggregateInput([
      key.count, plaintext.count, keyIdentifier.utf8.count,
      agreementPartyUInfo.count, agreementPartyVInfo.count,
      type.utf8.count, contentType.utf8.count,
    ])
    var operation = ReallyMeProtoJoseJweEncryptRequest()
    operation.keyManagementAlgorithm = protoKeyManagementAlgorithm(keyManagementAlgorithm)
    operation.contentEncryptionAlgorithm = protoContentEncryptionAlgorithm(
      contentEncryptionAlgorithm)
    operation.key = Data(key)
    operation.plaintext = Data(plaintext)
    operation.kid = keyIdentifier
    operation.apu = Data(agreementPartyUInfo)
    operation.apv = Data(agreementPartyVInfo)
    operation.typ = type
    operation.cty = contentType
    var request = ReallyMeProtoJoseOperationRequest()
    request.operation = .jweEncrypt(operation)
    let response = try execute(&request)
    guard case .jweEncrypt(let selected)? = response.response,
      selected.unknownFields.data.isEmpty
    else {
      throw ReallyMeJOSEError.malformedProviderResponse
    }
    switch selected.outcome {
    case .result(let result):
      guard result.unknownFields.data.isEmpty else {
        throw ReallyMeJOSEError.malformedProviderResponse
      }
      return result.compact
    case .error(let error): throw try sdkError(error)
    case nil: throw ReallyMeJOSEError.malformedProviderResponse
    }
  }

  public func decryptJWE(
    compact: String,
    keyManagementAlgorithm: ReallyMeJOSEJWEKeyManagementAlgorithm,
    contentEncryptionAlgorithm: ReallyMeJOSEJWEContentEncryptionAlgorithm,
    key: [UInt8],
    headerPolicy: ReallyMeJOSEJWEHeaderPolicy? = nil
  ) throws -> [UInt8] {
    try requireAggregateInput([compact.utf8.count, key.count])
    var operation = ReallyMeProtoJoseJweDecryptRequest()
    operation.compact = compact
    operation.keyManagementAlgorithm = protoKeyManagementAlgorithm(keyManagementAlgorithm)
    operation.contentEncryptionAlgorithm = protoContentEncryptionAlgorithm(
      contentEncryptionAlgorithm)
    operation.key = Data(key)
    if let headerPolicy { operation.headerPolicy = protoJWEHeaderPolicy(headerPolicy) }
    var request = ReallyMeProtoJoseOperationRequest()
    request.operation = .jweDecrypt(operation)
    var response = try execute(&request)
    guard case .jweDecrypt(var selected)? = response.response,
      selected.unknownFields.data.isEmpty
    else {
      throw ReallyMeJOSEError.malformedProviderResponse
    }
    response.response = nil
    switch selected.outcome {
    case .result(var result):
      guard result.unknownFields.data.isEmpty else {
        throw ReallyMeJOSEError.malformedProviderResponse
      }
      selected.outcome = nil
      defer { ReallyMeJOSEMemory.clearOwned(&result.plaintext) }
      return [UInt8](result.plaintext)
    case .error(let error): throw try sdkError(error)
    case nil: throw ReallyMeJOSEError.malformedProviderResponse
    }
  }

  /// Explicit wire-level API. The caller owns and must clear returned bytes.
  public func executeWireRequest(_ request: [UInt8]) throws -> [UInt8] {
    try provider.executeBinary(request)
  }

  /// Explicit generated-ProtoJSON request API returning canonical binary response bytes.
  public func executeWireJSONRequest(_ request: [UInt8]) throws -> [UInt8] {
    try provider.executeJSON(request)
  }

  private func execute(
    _ request: inout ReallyMeProtoJoseOperationRequest
  ) throws -> ReallyMeProtoJoseOperationResponse {
    defer { wipeRequest(&request) }
    var requestBytes: [UInt8]
    do {
      requestBytes = try request.serializedBytes()
    } catch {
      throw ReallyMeJOSEError.malformedProviderResponse
    }
    defer { provider.clearOwned(&requestBytes) }
    var responseBytes = try provider.executeBinary(requestBytes)
    defer { provider.clearOwned(&responseBytes) }
    let response: ReallyMeProtoJoseOperationResponse
    do {
      var options = BinaryDecodingOptions()
      options.messageDepthLimit = 32
      response = try ReallyMeProtoJoseOperationResponse(
        serializedBytes: responseBytes,
        options: options
      )
    } catch {
      throw ReallyMeJOSEError.malformedProviderResponse
    }
    guard response.contractVersion == .v1,
      response.unknownFields.data.isEmpty
    else {
      throw ReallyMeJOSEError.malformedProviderResponse
    }
    if case .boundaryError(let error)? = response.response {
      throw try sdkError(error)
    }
    return response
  }

  private func requireAggregateInput(_ lengths: [Int]) throws {
    var aggregate = 0
    for length in lengths {
      let (sum, overflow) = aggregate.addingReportingOverflow(length)
      guard !overflow, sum <= provider.maximumBinaryRequestBytes else {
        throw ReallyMeJOSEError.jose(
          branch: .primitive,
          reason: .commonResourceLimitExceeded
        )
      }
      aggregate = sum
    }
  }
}

private func sdkError(_ error: ReallyMeProtoJoseError) throws -> ReallyMeJOSEError {
  guard error.unknownFields.data.isEmpty else {
    throw ReallyMeJOSEError.malformedProviderResponse
  }
  let branch: ReallyMeJOSEErrorBranch
  let protoReason: ReallyMeProtoJoseErrorReason
  switch error.error {
  case .primitive(let value):
    guard value.unknownFields.data.isEmpty else {
      throw ReallyMeJOSEError.malformedProviderResponse
    }
    branch = .primitive
    protoReason = value.reason
  case .provider(let value):
    guard value.unknownFields.data.isEmpty else {
      throw ReallyMeJOSEError.malformedProviderResponse
    }
    branch = .provider
    protoReason = value.reason
  case .backend(let value):
    guard value.unknownFields.data.isEmpty else {
      throw ReallyMeJOSEError.malformedProviderResponse
    }
    branch = .backend
    protoReason = value.reason
  case nil:
    throw ReallyMeJOSEError.malformedProviderResponse
  }
  guard let reason = ReallyMeJOSEErrorReason(rawValue: protoReason.rawValue) else {
    throw ReallyMeJOSEError.malformedProviderResponse
  }
  return .jose(branch: branch, reason: reason)
}

private func protoSignatureAlgorithm(
  _ value: ReallyMeJOSESignatureAlgorithm
) -> ReallyMeProtoJoseSignatureAlgorithm {
  switch value {
  case .edDSA: .eddsa
  case .es256: .es256
  }
}

private func protoKeyManagementAlgorithm(
  _ value: ReallyMeJOSEJWEKeyManagementAlgorithm
) -> ReallyMeProtoJoseJweKeyManagementAlgorithm {
  switch value {
  case .direct: .direct
  case .ecdhESP256: .ecdhEsP256
  case .ecdhESP384: .ecdhEsP384
  case .ecdhESP521: .ecdhEsP521
  }
}

private func protoContentEncryptionAlgorithm(
  _ value: ReallyMeJOSEJWEContentEncryptionAlgorithm
) -> ReallyMeProtoJoseJweContentEncryptionAlgorithm {
  switch value {
  case .a128GCM: .a128Gcm
  case .a192GCM: .a192Gcm
  case .a256GCM: .a256Gcm
  }
}

private func protoJWTHeaderPolicy(
  _ value: ReallyMeJOSEJWTHeaderPolicy
) -> ReallyMeProtoJoseJwtHeaderValidationPolicy {
  var result = ReallyMeProtoJoseJwtHeaderValidationPolicy()
  result.allowMissingTyp = value.allowMissingTyp
  result.allowEmbeddedKeyHeader = value.allowEmbeddedKeyHeader
  result.acceptedTypValues = value.acceptedTypValues
  return result
}

private func protoJWTTemporalPolicy(
  _ value: ReallyMeJOSEJWTTemporalPolicy
) -> ReallyMeProtoJoseJwtTemporalValidationPolicy {
  var result = ReallyMeProtoJoseJwtTemporalValidationPolicy()
  result.requireExp = value.requireExpiration
  result.requireNbf = value.requireNotBefore
  result.requireIat = value.requireIssuedAt
  result.clockSkewSeconds = value.clockSkewSeconds
  result.maxFutureIatSkewSeconds = value.maximumFutureIssuedAtSkewSeconds
  result.nowUnix = value.nowUnix
  result.expectedAudience = value.expectedAudience
  result.expectedIssuer = value.expectedIssuer ?? ""
  result.expectedSubject = value.expectedSubject ?? ""
  return result
}

private func protoJWEHeaderPolicy(
  _ value: ReallyMeJOSEJWEHeaderPolicy
) -> ReallyMeProtoJoseJweHeaderValidationPolicy {
  var result = ReallyMeProtoJoseJweHeaderValidationPolicy()
  result.requireKid = value.requireKeyIdentifier
  if let expected = value.expectedKeyIdentifier {
    var wrapped = ReallyMeProtoJoseExpectedString()
    wrapped.value = expected
    result.expectedKid = wrapped
  }
  if let expected = value.expectedType {
    var wrapped = ReallyMeProtoJoseExpectedString()
    wrapped.value = expected
    result.expectedTyp = wrapped
  }
  if let expected = value.expectedContentType {
    var wrapped = ReallyMeProtoJoseExpectedString()
    wrapped.value = expected
    result.expectedCty = wrapped
  }
  if let expected = value.expectedAgreementPartyUInfo {
    var wrapped = ReallyMeProtoJoseExpectedBytes()
    wrapped.value = Data(expected)
    result.expectedApu = wrapped
  }
  if let expected = value.expectedAgreementPartyVInfo {
    var wrapped = ReallyMeProtoJoseExpectedBytes()
    wrapped.value = Data(expected)
    result.expectedApv = wrapped
  }
  return result
}

private func wipeRequest(_ request: inout ReallyMeProtoJoseOperationRequest) {
  switch request.operation {
  case .jwsSign:
    ReallyMeJOSEMemory.clearOwned(&request.jwsSign.privateKey)
    ReallyMeJOSEMemory.clearOwned(&request.jwsSign.payload)
  case .jwsVerify:
    ReallyMeJOSEMemory.clearOwned(&request.jwsVerify.publicKey)
  case .jwtEncodeUnsigned:
    ReallyMeJOSEMemory.clearOwned(&request.jwtEncodeUnsigned.claimsJson)
  case .jwtDecodeUnsigned:
    break
  case .jwtSign:
    ReallyMeJOSEMemory.clearOwned(&request.jwtSign.claimsJson)
    ReallyMeJOSEMemory.clearOwned(&request.jwtSign.jwkJson)
    ReallyMeJOSEMemory.clearOwned(&request.jwtSign.privateKey)
  case .jwtVerify:
    ReallyMeJOSEMemory.clearOwned(&request.jwtVerify.jwkJson)
    ReallyMeJOSEMemory.clearOwned(&request.jwtVerify.publicKey)
  case .jweEncrypt:
    ReallyMeJOSEMemory.clearOwned(&request.jweEncrypt.key)
    ReallyMeJOSEMemory.clearOwned(&request.jweEncrypt.plaintext)
    ReallyMeJOSEMemory.clearOwned(&request.jweEncrypt.apu)
    ReallyMeJOSEMemory.clearOwned(&request.jweEncrypt.apv)
  case .jweDecrypt:
    ReallyMeJOSEMemory.clearOwned(&request.jweDecrypt.key)
  case nil:
    break
  }
  request.operation = nil
}
